### ISSUE TITLE
Conda 4.2.x attempts to create lock files in readonly file-based channels and does not respect CONDA_ENVS_PATH

### ISSUE THREAD

**Issue by gmnash** [conda/conda#3469]
I am getting warnings and eventually an error after upgrading from 4.1.11 to 4.2.7 relating to lock files and packages attempting to write to shared read-only directories in our continuous build.

I modified `/shared/path/to/miniconda/lib/python2.7/site-packages/conda/lock.py` to find out where conda 4.2.x was putting lock files and found the following:

`WARNING conda.lock:touch(55): Failed to create lock, do not run conda in parallel processes [errno 13] /shared/path/to/vendor/conda/linux-64.pid24915.conda_lock`

`WARNING conda.lock:touch(55): Failed to create lock, do not run conda in parallel processes [errno 13] /shared/path/to/miniconda/pkgs/python-dateutil-2.4.2-py27_0.tar.bz2.pid24915.conda_lock`

`/shared/path/to` is read-only because it is used by multiple build-agents concurrently.

Additional I encountered the following error:

`CondaRuntimeError: Runtime error: RuntimeError: Runtime error: Could not open u'/shared/path/to/miniconda/pkgs/python-dateutil-2.4.2-py27_0.tar.bz2.part' for writing ([Errno 13] Permission denied: u'/shared/path/to/miniconda/pkgs/python-dateutil-2.4.2-py27_0.tar.bz2.part').`

In conda <= 4.1.11 we were able to set the `CONDA_ENVS_PATH` to have a shared installation of conda, but with separate package caches/lock files per build agent on our continuous build server, in order to avoid concurrent use of the package cache.

Two questions:
* How do we disable lock files being placed in read-only shared file-based channels.
* Does `CONDA_ENVS_PATH` no longer override the package cache/lock file directory from the root?

**Comment by aldanor** [comment]
Re: lock files, ran into the same problem today -- conda tries to create lock files for readonly file-based channels which yields a ton of warnings:

```sh
$ conda create -p ./foobar python=3.5
WARNING conda.lock:touch(53): Failed to create lock, do not run conda in parallel processes [errno 13]
WARNING conda.lock:touch(53): Failed to create lock, do not run conda in parallel processes [errno 13]
WARNING conda.lock:touch(53): Failed to create lock, do not run conda in parallel processes [errno 13]
```

This is a regression from 4.1.x, please fix.

**Comment by aldanor** [comment]
@HugoTian @kalefranz ping

**Comment by aldanor** [comment]
By the way, why does it try to lock at all when all we're doing is reading from an external channel? I get that package cache lock makes sense, but why this?

**Comment by aldanor** [comment]
Anyone?..

**Comment by kcowgill** [comment]
We're bumping into this one as well. Has anyone got any info yet?

**Comment by gnarlie** [comment]
This is blocking our upgrade to 4.2 as well.

**Comment by eyew** [comment]
This issue with creating locks in read-only filesystem channels is a breaking our workflow as well

**Comment by aldanor** [comment]
@kalefranz :panda_face: :panda_face: :panda_face: :panda_face: :panda_face: :panda_face: :panda_face: :panda_face: :panda_face: :panda_face:

**Comment by HugoTian** [comment]
@aldanor , I just check the history, previously, failing create lock will be handled with debug level logging . So there is no warning info. 

I think there may be 2 solutions, first we change it back to debug, and there will be no warning. Second, we give another detailed warning based on exception types.

@gnarlie , @eyew , can you give us more info about the exception or error that you encountered during upgrade or workflow ?

Thanks in advance

**Comment by gmnash** [comment]
@HugoTian why does Conda attempt to put a lockfile in a file-based channel in the first place?  Are you saying all clients of a file-based channel need write access?  How does that work with http channels?

**Comment by eyew** [comment]
@HugoTian The error message I get is: Runtime error: Could not open u/path_to_package/pkgs/packagename.tar.bz2.part' for writing ([Errno 13] Permission denied: ...)

The  pkgs directory (and miniconda installation itself) is on a shared filesystem and read-only because that conda instance is shared among multiple continuous build agents. In conda 4.1.x each build agent can maintain its own (temporary) package cache and locks by setting `CONDA_ENVS_PATH`.

If we're going about this incorrectly, what is the proper way to share a conda installation among many agents? Having them each install miniconda to a temporary location before running a build is too slow to be workable in our case.

**Comment by aldanor** [comment]
There's two separate issues discussed here:

1) locking `pkgs` directory which generally is a legit thing to do and which may require further discussion
2) locking remote file channels -- which is a very strange thing to do, anyone care to explain?

**Comment by gmnash** [comment]
1) Locking the package directory makes sense.  However, in versions < 4.2.x, one could specify where the package directory was located via CONDA_ENV_PATH.  This no longer seems to work as of 4.2.x
2) Agreed.

**Comment by HugoTian** [comment]
I think we may bring back the CONDA_ENV_PATH setting, and I am not quite familiar with the cases when dealing with remote file-based channels. 

When I design the filelock and directory lock, the primary goal is to lock local file and directory to avoid parallel execution. Sorry that I did not consider the remote file channel case.

**Comment by mattgodbolt** [comment]
This has affected me too. I install to a known location using `-p` so even having to set the `CONDA_ENV_PATH` to a writable location is counter-intuitive. However, it's the only workaround I have, and in 4.2 it's gone too.

**Comment by ocarino** [comment]
Hi there,

we seem to run into the same issue. How is the CONDA_ENV_PATH variable supposed to work? I set "env_dirs" in .condarc to two values:

envs_dirs:
  - ~/conda-envs
  - /location/of/conda/installation/envs

This works so far as expected (so users installing something get this installed into their home directories (~/conda-envs)). But obviously some locking is still done in the installation dir of conda (which is not writeable for the users).

Would setting "CONDA_ENV_PATH" to some directory solve this issue (if it worked for 4.2.0)?

Cheers, Olaf

**Comment by kalefranz** [comment]
Sorry for the delayed response here.  With 4.2 you could set a `CONDA_ENVS_DIRS` environment variable.  `envs_dirs` is what the equivalent is named in the condarc file.  I missed the environment variable though I guess.

With #3685 the functionality of `CONDA_ENVS_PATH` should be fully restored.  Basically, `envs_path` is now an alias for `envs_dirs`.

**Comment by aldanor** [comment]
Are there any updates on the lock file problems?

**Comment by kalefranz** [comment]
The lock file issue is part of our challenge with getting multi-user installs right. It will be fixed within the next two months and ten days, i.e. within 2016.

**Comment by gmnash** [comment]
If we just use CONDA_ENVS_DIRS does that restore previous functionality
without a new release?  It seems from this comment that the environment
variable was just renamed.

**Comment by kalefranz** [comment]
Yes, it will restore previous functionality. In the latest conda 4.2 release either environment variable should work.

The comment here also relates to the lock file issue:
https://github.com/conda/conda/issues/3614#issuecomment-255923255

**Comment by smontanaro** [comment]
Was this fixed? If so, in what version? We are running 4.2.13 and getting these warnings.

**Comment by kalefranz** [comment]
This is fixed in conda 4.3.  It's in canary now, but we'll be releasing it soon.  If you want to use the latest canary versions,

    conda config --add channels conda-canary
    conda update conda
